org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/SymbolUtils.java

```diff
@@ -0,0 +1,159 @@
+				try {
+					IType type = (IType) element;
+					if (type.isInterface()) {
+						return SymbolKind.Interface;
+					} else if (type.isEnum()) {
+						return SymbolKind.Enum;
+					}
+				} catch (JavaModelException ignore) {
+				}
+				return SymbolKind.Class;
+			case IJavaElement.ANNOTATION:
+				return SymbolKind.Property; // TODO: find a better mapping
+			case IJavaElement.CLASS_FILE:
+			case IJavaElement.COMPILATION_UNIT:
+				return SymbolKind.File;
+			case IJavaElement.FIELD:
+				IField field = (IField) element;
+				try {
+					if (field.isEnumConstant()) {
+						return SymbolKind.EnumMember;
+					}
+					int flags = field.getFlags();
+					if (Flags.isStatic(flags) && Flags.isFinal(flags)) {
+						return SymbolKind.Constant;
+					}
+				} catch (JavaModelException ignore) {
+				}
+				return SymbolKind.Field;
+			case IJavaElement.IMPORT_CONTAINER:
+			case IJavaElement.IMPORT_DECLARATION:
+				//should we return SymbolKind.Namespace?
+			case IJavaElement.JAVA_MODULE:
+				return SymbolKind.Module;
+			case IJavaElement.INITIALIZER:
+				return SymbolKind.Constructor;
+			case IJavaElement.LOCAL_VARIABLE:
+				return SymbolKind.Variable;
+			case IJavaElement.TYPE_PARAMETER:
+				return SymbolKind.TypeParameter;
+			case IJavaElement.METHOD:
+				try {
+					// TODO handle `IInitializer`. What should be the `SymbolKind`?
+					if (element instanceof IMethod) {
+						if (((IMethod) element).isConstructor()) {
+							return SymbolKind.Constructor;
+						}
+					}
+					return SymbolKind.Method;
+				} catch (JavaModelException e) {
+					return SymbolKind.Method;
+				}
+			case IJavaElement.PACKAGE_DECLARATION:
+				return SymbolKind.Package;
+		}
+		return SymbolKind.String;
+	}
+
+	static String getName(IJavaElement element) {
+		String name = JavaElementLabelsCore.getElementLabel(element, ALL_DEFAULT);
+		return name == null ? element.getElementName() : name;
+	}
+
+	static Range getRange(IJavaElement element) throws JavaModelException {
+		Location location = JDTUtils.toLocation(element, FULL_RANGE);
+		return location == null ? DEFAULT_RANGE : location.getRange();
+	}
+
+	static Range getSelectionRange(IJavaElement element) throws JavaModelException {
+		Location location = JDTUtils.toLocation(element);
+		return location == null ? DEFAULT_RANGE : location.getRange();
+	}
+
+	static String getDetail(IJavaElement element, String name) {
+		return constructDetail(element, name, ALL_DEFAULT | M_APP_RETURNTYPE | ROOT_VARIABLE);
+	}
+
+	static String getDetail(IJavaElement element, String name, long additionalFlags) {
+		return constructDetail(element, name, ALL_DEFAULT | ROOT_VARIABLE | additionalFlags);
+	}
+
+	private static String constructDetail(IJavaElement element, String name, long flags) {
+		String nameWithDetails = JavaElementLabelsCore.getElementLabel(element, flags);
+		if (nameWithDetails != null && nameWithDetails.startsWith(name)) {
+			return nameWithDetails.substring(name.length()).trim();
```

Comment:
The `trim()` here is messing with some tests, need to figure out if it really is needed.